Adding Setting Aggregate function input format to allow Insert queries into tables with AggregateFunction columns

punithns97 opened the pull request:
Resolves #87827.

<!--- Disable AI PR formatting assistant: true -->
### Changelog category (leave one):
- Improvement

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Adds session-level setting `aggregate_function_input_format` to improve `INSERT` queries into tables with AggregateFunction columns, allowing insertion of data as serialized state, raw values, or arrays. 

### Documentation entry for user-facing changes
Adds a Session level setting `aggregate_function_input_format` with the following possible values:
- `state` - binary string with the serialized state (the default)
- `value` - the format will expect a single value of the argument of the aggregate function, or in the case of multiple arguments, a tuple of them; that will be deserialized to form the relevant state
- `array` - the format will expect an Array of values, as described in the values option above; all the elements of the array will be aggregated to form the state

### Details
The goal of this PR is to allow the usage of `AggregateFunction` to support various other formats like 'JSON', 'CSV', 'TSV'.

Resolves #87827.

#### Example use

For a table with this structure:
```sql
CREATE TABLE test_agg_single (
    user_id UInt64,
    avg_session_length AggregateFunction(avg, UInt32)
)
```

The user can `SET aggregate_function_input_format = 'value'` and perform queries such as:
```sql
INSERT INTO test_agg_single VALUES (124, '456'), (125, '789'), (126, '321');
```

Or the user can `SET aggregate_function_input_format = 'array'`:
```sql
INSERT INTO test_agg_single VALUES (127, '[100,200,300]'), (128, '[400,500]'), (129, '[600]');
```

punithns97 committed:
Added Settings for Aggregate Input Function Format

punithns97 committed:
Added Setting to Setting History

punithns97 committed:
Added parsing logic for deserializing for Values and Array

punithns97 committed:
Added function to deserialize based on input

punithns97 committed:
Added flow to parse aggregate input function format

punithns97 committed:
Update FormatSettings

punithns97 committed:
Adding test cases

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=88088) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=88088) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=88088) it.</sub>

punithns97 commented:
Same as https://github.com/ClickHouse/ClickHouse/pull/88049 . 
Raising new PR cause I changed accounts.

punithns97 committed:
Formatting Documentation

punithns97 committed:
Merge branch 'ClickHouse:master' into master

punithns97 committed:
Update SerializationAggregateFunction.cpp

punithns97 committed:
Merge branch 'ClickHouse:master' into master

punithns97 committed:
Making test file executable

punithns97 committed:
Fixing test cases

punithns97 committed:
Fixing Style Check Errors

punithns97 committed:
Fixing Style Check

punithns97 committed:
Fixing Style Check

punithns97 committed:
Fixing Style Check

punithns97 committed:
Fixing Test Cases

punithns97 committed:
Merge branch 'ClickHouse:master' into master

punithns97 committed:
Fixing Test Cases

punithns97 committed:
Merge branch 'ClickHouse:master' into master

punithns97 committed:
Merge branch 'ClickHouse:master' into master

punithns97 committed:
Merge branch 'ClickHouse:master' into master

GrigoryPervakov reviewed (commented):
I'll check the full PR a bit later, now read only some parts

punithns97 committed:
Merge branch 'ClickHouse:master' into master

punithns97 committed:
Merge branch 'ClickHouse:master' into master

punithns97 committed:
Refactoring Aggregate Function Input Format Setting as Enum

punithns97 committed:
Registering Setting

punithns97 committed:
Fixing Style Check

punithns97 committed:
Merge branch 'ClickHouse:master' into master

punithns97 commented on src/Core/Settings.cpp:
```diff
@@ -4640,6 +4640,41 @@ With `aggregate_functions_null_for_empty = 1` the result would be:
+
+Possible values:
+
+- `state` — Binary string with the serialized state (the default). This is the current behavior where AggregateFunction values are expected as binary data.
```
Updated in Latest Commits.

punithns97 commented on tests/queries/0_stateless/03645_aggregate_function_input_format.sh:
```diff
@@ -0,0 +1,208 @@
+#!/usr/bin/env bash
```
I wrote it as a .sh script because parsing inputs in formats like JSON and CSV isn’t straightforward in SQL, and I also wanted to test cases both with and without using the VALUES function to parse input. Using a shell script made it easier to handle these varied formats and input methods as well as Error Handling scenarios.

punithns97 committed:
Using Inbuilt Serializer

punithns97 committed:
Merge branch 'master' into master

punithns97 committed:
Merge branch 'master' into master

punithns97 committed:
Merge branch 'ClickHouse:master' into master

punithns97 commented on src/DataTypes/Serializations/SerializationAggregateFunction.cpp:
```diff
@@ -127,23 +133,163 @@ static void deserializeFromString(const AggregateFunctionPtr & function, IColumn
+        const auto & argument_types = function->getArgumentTypes();
+        // Parse the array - expect format like [val1,val2,val3] or [(val1a,val1b),(val2a,val2b)]
+        ReadBufferFromString buf(array_str);
+        if (buf.eof() || *buf.position() != '[')
```
@GrigoryPervakov  Do you suggest that I remove this manual deserializer in the deserializeFromArray function and instead use the existing deserializer? This current implementation using the manual deserializer explicitly handles multiple cases—such as JsonString, numeric arrays, and string arrays (with double quotes) while switching to the existing deserializer would lose some of that flexibility.

GrigoryPervakov commented on src/DataTypes/Serializations/SerializationAggregateFunction.cpp:
```diff
@@ -127,23 +133,163 @@ static void deserializeFromString(const AggregateFunctionPtr & function, IColumn
+        if (buf.eof() || *buf.position() != ']')
+            throw Exception(ErrorCodes::CANNOT_PARSE_INPUT_ASSERTION_FAILED,
+                "Expected closing bracket in array, got: '{}'", array_str);
+    }
```
I have not tested it, but for me, it's a bit more readable parsing logic, and it also does not recreate columns and serialization for every value
```suggestion
     try
    {
        // Get the argument types for the aggregate function
        const auto & argument_types = function->getArgumentTypes();

        auto elem_type= argument_types.size() == 1 ? argument_types[0] : std::make_shared<DataTypeTuple>(argument_types);
        auto tmp_column = elem_type->createColumn();
        auto elem_serialization = elem_type->getDefaultSerialization();
        absl::InlinedVector<const IColumn *, 7> columns_ptrs;
        if (argument_types.size() == 1)
            columns_ptrs.push_back(tmp_column.get());
        else
            for (const auto & col : assert_cast<const ColumnTuple*>(tmp_column.get())->getColumns())
                columns_ptrs.push_back(col.get());

        // Parse the array - expect format like [val1,val2,val3] or [(val1a,val1b),(val2a,val2b)]
        ReadBufferFromString buf(array_str);
        assertChar('[', buf);
        bool first = true;
        while (!buf.eof())
        {
            skipWhitespaceIfAny(buf);
            if (*buf.position() == ']')
                break;

            if (!first)
            {
                assertChar(',', buf);
                skipWhitespaceIfAny(buf);
            }
            first = false;

            elem_serialization->deserializeTextQuoted(*tmp_column, buf, settings);
            function->add(place, columns_ptrs.data(), 0, &arena);
            tmp_column->popBack(1);
        }
        assertChar(']', buf);
    }
```

GrigoryPervakov reviewed (commented):
> Please review the PR and provide any comments or feedback on the changes.

@punithns97  It's overall LGTM, added one more proposal, and the settings history should be fixed.

Sorry for the long delay

punithns97 committed:
Merge branch 'ClickHouse:master' into master

punithns97 committed:
Refactoring Code and Settings History

punithns97 committed:
Fixing Style Check

punithns97 committed:
Remove unused error code from SerializationAggregateFunction

punithns97 committed:
Changing string function param to pass by const reference

punithns97 committed:
Merge branch 'master' into master

punithns97 committed:
Update description for aggregate_function_input_format setting

punithns97 committed:
Move aggregate_function_input_format setting to 25.11

Reintroduce aggregate_function_input_format setting to control input format during INSERT operations.

punithns97 committed:
Add aggregate_function_input_format setting to 25.12

punithns97 committed:
Merge branch 'master' into master

GrigoryPervakov committed:
add docs

GrigoryPervakov committed:
Merge branch 'master' into master

punithns97 committed:
Merge branch 'ClickHouse:master' into master

punithns97 committed:
Merge branch 'ClickHouse:master' into master

punithns97 commented:
@GrigoryPervakov  The checks which are failing are not related to my changes. The tests seem to be flaky as mentioned [here](https://github.com/ClickHouse/ClickHouse/issues/90721) .

GrigoryPervakov committed:
Merge branch 'master' into master

punithns97 committed:
Merge branch 'ClickHouse:master' into master

punithns97 committed:
Merge branch 'ClickHouse:master' into master

punithns97 committed:
Merge branch 'ClickHouse:master' into master

punithns97 commented:
> It should work for all input formats, including RowBinary. If it's not the case, please revert this PR and re-implement it.
> I can argue that this is especially important for the RowBinary format, and having a partial implementation misses the point.

This works with only text-based formats. for supporting binary format, I can raise another PR to supplement this ?

clickgapai commented:
Hi @punithns97 @GrigoryPervakov — while reviewing this PR I found the following:

- **Bug (P0):** `deserializeFromValue` single-arg path uses `deserializeTextCSV` on already-extracted value, causing silent truncation at comma for String types — https://github.com/ClickHouse/ClickHouse/issues/103033

Happy to discuss — close anything that's wrong or already addressed.
